---
title: "Synthetic Single-Cell RNA-Sequencing Data from Small Pilot Studies using Deep Generative Models"
subtitle: "Figure S4"
author:
- affiliation: Institute of Medical Biometry and Statistics, Faculty of Medicine and
    Medical Center - University of Freiburg
  name: "Martin Treppner"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    fig_width: 15
    fig_height: 8
    theme: united
    toc: yes
  pdf_document:
    toc: yes
---

# Loading required packages  

```{r, warning=FALSE, message=FALSE}
options(repos="https://CRAN.R-project.org")
if (!require("pacman")) install.packages("pacman")
pacman::p_load(dplyr,cowplot,ggplot2,tidyverse,gt)
```

# Theme  

```{r, warning=FALSE, message=FALSE}
dviz_font_family <- ""
theme_dviz_open <- function(font_size = 14, font_family = dviz_font_family, line_size = .5,
                          rel_small = 12/14, rel_tiny = 11/14, rel_large = 16/14) {
  half_line <- font_size / 2
  
  cowplot::theme_half_open(font_size = font_size, font_family = font_family, line_size = line_size,
                           rel_small = rel_small, rel_tiny = rel_tiny, rel_large = rel_large)  %+replace%
    theme(
      plot.margin = margin(half_line/2, 1.5, half_line/2, 1.5),
      complete = TRUE
    )
}
```

# Read original data and gene names  

```{r, warning=FALSE, message=FALSE}
orig_dat <- read.csv("/Users/martintreppner/Desktop/notebook_test/segerstolpe_hvg.csv")
orig_dat <- orig_dat[,2:ncol(orig_dat)] %>% t() %>% as.data.frame()
gene_names <- read.csv("/Users/martintreppner/Desktop/notebook_test/segerstolpe_gene_names.csv")
gene_names <- gene_names[,2]
colnames(orig_dat) <- gene_names
```

# Read synthetic data  

```{r, warning=FALSE, message=FALSE}
scdbm_dat <- read.csv("/Users/martintreppner/Desktop/notebook_test/gen_data_scdbm384_seed24.csv")
scvi_dat <- read.csv("/Users/martintreppner/Desktop/notebook_test/gen_data_scvi384_seed44.csv")

colnames(scdbm_dat) <- gene_names
colnames(scvi_dat) <- gene_names
```

# Plot exemplary gene expression distributions  

```{r, warning=FALSE, message=FALSE}
genelist <- c("PRSS1", "KRT19", "TIMP1", "PAX6", "PLVAP", "GCG", "INS", "CTRB1", "NEUROD1", "MAFB", "PDX1", "CASR")

orig_dat$Model <- "Original"
scdbm_dat$Model <- "scDBM"
combined1 <- rbind(orig_dat, scdbm_dat)

combined1$Model <- factor(combined1$Model, c("scDBM", "Original"))

combined1 <- combined1[,c(genelist, "Model")]
combined1$Col <- "scDBM"

scvi_dat$Model <- "scVI"
combined2 <- rbind(orig_dat, scvi_dat)

combined2$Model <- factor(combined2$Model, c("scVI", "Original"))

combined2 <- combined2[,c(genelist, "Model")]
combined2$Col <- "scVI"

complete_plot <- rbind(combined1,combined2)

complete_plot_gather <- complete_plot %>% gather(genelist, key = "Gene", value = "Expression")

complete_plot_gather$Expression <- log1p(complete_plot_gather$Expression + 1)

complete_plot_gather$Model <- factor(complete_plot_gather$Model, levels = c("Original", "scDBM", "scVI"))
complete_plot_gather$Col <- factor(complete_plot_gather$Col, levels = c("scDBM", "scVI"))

p <- ggplot(complete_plot_gather, aes(x = Gene, y = Expression, col = Model)) +
  geom_violin(fill = 'grey90') +
  facet_wrap(~Gene, ncol=3,scales="free") +
  xlab("") + 
  ylab("Log(Read + 1)") +
  scale_colour_manual(values = c("#D55E00", "#0072B2", "#009E73")) +
  theme_dviz_open() +
  theme(plot.margin = unit(c(0.25,0.45,0.25,0.45), "cm"),
        strip.background = element_blank(),
        strip.text.x = element_blank(),
        legend.position="none",
        legend.title=element_blank(),
        legend.justification = "center")
p
```

# Session info  

```{r}
sessionInfo()
```



