---
title: "Synthetic Single-Cell RNA-Sequencing Data from Small Pilot Studies using Deep Generative Models"
subtitle: "Figure 2"
author:
- affiliation: Institute of Medical Biometry and Statistics, Faculty of Medicine and
    Medical Center - University of Freiburg
  name: "Martin Treppner"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    fig_width: 15
    fig_height: 8
    theme: united
    toc: yes
  pdf_document:
    toc: yes
---

# Loading required packages  

```{r, warning=FALSE, message=FALSE}
options(repos="https://CRAN.R-project.org")
if (!require("pacman")) install.packages("pacman")
pacman::p_load(dplyr,cowplot,ggplot2,tidyverse,gt)
```

# Davies-Bouldin index for clustering  

```{r}
dbi_dat <- read.csv("/Users/martintreppner/Desktop/notebook_test/dbi_plotting.csv", header = TRUE)
cutoff <- dplyr::filter(dbi_dat, variable == "Original")$value[1]
dbi_dat <- rbind(dplyr::filter(dbi_dat, variable == c("scDBM")),dplyr::filter(dbi_dat, variable == c("scVI")))
```

```{r, warning=FALSE, message=FALSE}
p <- ggplot(dbi_dat, aes(x=factor(cells, levels = c("384", "768", "1152", "1536", "1920")), y=value, fill=factor(variable, levels = c("scDBM", "scVI")))) +
      theme_minimal_hgrid(font_size = 25) +
      scale_y_continuous(
        limits = c(0, 15), expand = c(0, 0),
        sec.axis = dup_axis(
        breaks = c(cutoff),
        labels = c("Original"),
        name = NULL,
        guide = guide_axis(angle = 0, n.dodge = 1.5))
      ) +
      scale_x_discrete(labels=c("1 Plate", "2 Plates", "3 Plates", "4 Plates", "5 Plates")) +
      scale_fill_discrete(c("#0072B2", "#009E73", "#5B3794"), labels = c("NB-DBM Upsample", "scVI Upsample", "scVI Posterior")) +
      geom_hline(yintercept = cutoff, colour = "#E69F00", lwd = 1.5, linetype = "dotted") +
      geom_boxplot(size = 0.4) +
      theme(legend.title=element_blank(), legend.position = "bottom", legend.justification = "center") +
      labs(y = expression(paste("Davies-Bouldin Index")), x = "")
p <- p + scale_fill_manual(values=c("#0072B2", "#009E73", "#5B3794"),labels = c("scDBM", "scVI"))
p
```

# Adjusted Rand index for clustering  

```{r, warning=FALSE, message=FALSE}
ari_dat <- read.csv("/Users/martintreppner/Desktop/notebook_test/ari_plotting.csv", header = TRUE)
```

```{r, warning=FALSE, message=FALSE}
p1 <- ggplot(ari_dat, aes(x=factor(cells, levels = c("384", "768", "1152", "1536", "1920")), y=ARI, fill=factor(model, levels = c("scDBM", "scVI")))) +
      geom_boxplot(size = 0.4) + 
      theme_minimal_grid(font_size = 40) +
      scale_y_continuous(limits = c(0, 1)) +
      scale_x_discrete(labels=c("1 Plate", "2 Plates", "3 Plates", "4 Plates", "5 Plates")) +
      scale_fill_discrete(c("#0072B2", "#009E73"), labels = c("scDBM", "scVI")) +
      theme(legend.title=element_blank(), legend.position = "bottom", legend.justification = "center") +
      labs(y = expression(paste("Adjusted Rand Index")), x = "")
p1 <- p1 + scale_fill_manual(values=c("#0072B2", "#009E73"),labels = c("scDBM", "scVI"))
p1
```

# Session info  

```{r}
sessionInfo()
```

