---
title: "Synthetic Single-Cell RNA-Sequencing Data from Small Pilot Studies using Deep Generative Models - Data Preparation"
author:
- name: Martin Treppner
  affiliation: Institute of Medical Biometry and Statistics, Faculty of Medicine and Medical Center - University of Freiburg
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
output:
  BiocStyle::html_document:
    number_sections: yes
    toc: true
    toc_depth: 2
    bibliography: bibliography.bib
---

# Loading data using Bioconductor  

```{r}
library(scRNAseq)
sce.seger <- SegerstolpePancreasData()
```

# Data annotation  

```{r}
library(AnnotationHub)
edb <- AnnotationHub()[["AH73881"]]
symbols <- rowData(sce.seger)$symbol
ens.id <- mapIds(edb, keys=symbols, keytype="SYMBOL", column="GENEID")
ens.id <- ifelse(is.na(ens.id), symbols, ens.id)
```

# Data cleaning  

```{r}
# Removing duplicated rows.
keep <- !duplicated(ens.id)
sce.seger <- sce.seger[keep,]
rownames(sce.seger) <- ens.id[keep]

emtab.meta <- colData(sce.seger)[,c("cell type", "disease",
                                    "individual", "single cell well quality")]
colnames(emtab.meta) <- c("CellType", "Disease", "Donor", "Quality")
colData(sce.seger) <- emtab.meta

sce.seger$CellType <- gsub(" cell", "", sce.seger$CellType)
sce.seger$CellType <- paste0(
  toupper(substr(sce.seger$CellType, 1, 1)),
  substring(sce.seger$CellType, 2))

unfiltered <- sce.seger

low.qual <- sce.seger$Quality == "low quality cell"
```

```{r}
library(scater)
stats <- perCellQCMetrics(sce.seger)
qc <- quickPerCellQC(stats, percent_subsets="altexps_ERCC_percent",
                     batch=sce.seger$Donor,
                     subset=!sce.seger$Donor %in% c("HP1504901", "HP1509101"))

sce.seger <- sce.seger[,!(qc$discard | low.qual)]

colData(unfiltered) <- cbind(colData(unfiltered), stats)
unfiltered$discard <- qc$discard
```

# Extract highly variable genes  

```{r}
library(scran)
clusters <- quickCluster(sce.seger)
sce.seger <- computeSumFactors(sce.seger, clusters=clusters)
sce.seger <- logNormCounts(sce.seger) 

for.hvg <- sce.seger[,librarySizeFactors(altExp(sce.seger)) > 0 & sce.seger$Donor!="AZ"]
dec.seger <- modelGeneVarWithSpikes(for.hvg, "ERCC", block=for.hvg$Donor)
chosen.hvgs <- getTopHVGs(dec.seger, n=2000)

seger_hvg <- sce.seger[chosen.hvgs,]
```

```{r}
dat <- counts(seger_hvg)
write.csv(dat, "/Users/martintreppner/Desktop/notebook_test/segerstolpe_hvg.csv")

gene_names <- rowData(seger_hvg)$symbol
write.csv(gene_names, "/Users/martintreppner/Desktop/notebook_test/segerstolpe_gene_names.csv")

clust_lavels <- as.numeric(as.factor(colData(seger_hvg)[,1]))
write.csv(clust_lavels, "/Users/martintreppner/Desktop/notebook_test/segerstolpe_hvg_clustering.csv")
```

# Session info  

```{r}
sessionInfo()
```

